[WIP] Add more dual stack tests

aojea opened the pull request:
Add e2e and integration tests for the new dual stack feature that pluralize cluster IPs

danwinship commented on test/integration/dualstack/dualstack_test.go:
```diff
@@ -0,0 +1,1386 @@
+	netutils "k8s.io/utils/net"
+)
+
+// TestCreateServiceSingleStackIPv4 test the Service dualstackness in an IPv4 SingleStack cluster
```
can't really test "dualstackness" in a single-stack cluster... "test the dual-stack Service APIs"?

danwinship commented on test/integration/dualstack/dualstack_test.go:
```diff
@@ -0,0 +1,1386 @@
+	client := clientset.NewForConfigOrDie(&restclient.Config{Host: s.URL})
+
+	// Wait until the default "kubernetes" service is created.
+	if err = wait.Poll(250*time.Millisecond, time.Minute, func() (bool, error) {
```
If this was an e2e test I'd say you shouldn't use hardcoded timeouts and should use some constant from `tests/e2e/framework/...` instead. I'm not sure about integration tests...

danwinship commented on test/integration/dualstack/dualstack_test.go:
```diff
@@ -0,0 +1,1386 @@
+			expectError:    false,
+		},
+		{
+			name:           "Type ClusterIP - Server Allocated IP - IPv4 Family - Policy Required Dual Stack",
```
"Require" (no "d")
(everywhere)

danwinship commented on test/integration/dualstack/dualstack_test.go:
```diff
@@ -0,0 +1,1386 @@
+	}
+}
+
+// TestCreateServiceDualStackIPv4Only test the Service dualstackness in an IPv4 only DualStack cluster
```
"in a cluster with the dual-stack feature gate enabled but an IPv4-only configuration"

danwinship commented on test/integration/dualstack/dualstack_test.go:
```diff
@@ -0,0 +1,1386 @@
+
+	for j, ip := range svc.Spec.ClusterIPs {
+		if ip == v1.ClusterIPNone {
+			// ???
```
I forget if we decided this is allowed or not, but if it is, it would be an error in these tests

danwinship reviewed (commented):
Good first draft. There is a ton of duplication here that can be reduced:

1. I'm pretty sure you only need a single "test-runner" function, that takes arguments indicating the configured CIDR(s), feature gate status, etc.
2. The set of test cases is identical in each case; it's only the expected results that differ. Instead of having a single `expectError`, you could have `singleStackIPv4OK`, `singleStackIPv6OK`, `dualStackOK` and then you'd only need one copy of the array.
3. (Related to point 2) "Single stack IPv4 with the feature gate disabled" and "single stack IPv4 with the feature gate enabled" have exactly the same expected results, and likewise for the two single-stack IPv6 versions. The two dual stack versions (v4v6 / v6v4) have different expected `IPFamilies`/`ClusterIPs`, but the same expected success/failure.

Commit:
dualstack endpoints integration tests

add integration tests to verify the behaviour of the endpoints
and endpointslices controller with dual stack services.

Since services can be single or dual stack, endpoints should be
generated for each IP family in the endpoint slice controller.

The legacy endpoint controller only will generate endpoints
in the first IP family configured in the service.

integration fix

Commit:
dual stack e2e test

Dual stack services have two ClusterIPs, we already have tests that
exercise the connectivity from different scenarios to the first
ClusterIP of the service.

This PR adds a new constructor to the e2e network utils to enable
DualStack services, and replicate the same tests but using the
secondary ClusterIP this time, so we cover the connectivity to
both cluster IPs.

aojea commented:
/test pull-kubernetes-e2e-kind-dual-canary

aojea commented:
/test pull-kubernetes-e2e-azure-dualstack
/test pull-kubernetes-e2e-iptables-azure-dualstack
/test pull-kubernetes-e2e-kind-ipvs-dual-canary
/test pull-kubernetes-e2e-kind-dual-canary

aojea commented:
/test pull-kubernetes-e2e-azure-dualstack
/test pull-kubernetes-e2e-iptables-azure-dualstack
/test pull-kubernetes-e2e-kind-ipvs-dual-canary
/test pull-kubernetes-e2e-kind-dual-canary
